cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/MapReduceProgramRunner.java

```diff
@@ -231,35 +240,46 @@ public void starting() {
           // If RunId is not time-based, use current time as start time
           startTimeInSeconds = TimeUnit.MILLISECONDS.toSeconds(System.currentTimeMillis());
         }
-        if (workflowName == null) {
+        if (workflowRunId == null) {
           store.setStart(program.getId(), runId.getId(), startTimeInSeconds, twillRunId,
                          userArgs.asMap(), arguments.asMap());
         } else {
           // Program started by Workflow
-          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowName, workflowRunId, workflowNodeId,
-                                        startTimeInSeconds, twillRunId);
+          store.setWorkflowProgramStart(program.getId(), runId.getId(), workflowRunId.getProgram(),
+                                        workflowRunId.getRun(), workflowNodeId, startTimeInSeconds, twillRunId);
         }
       }
 
       @Override
       public void terminated(Service.State from) {
         closeAllQuietly(closeables);
+        ProgramRunStatus runStatus = ProgramController.State.COMPLETED.getRunStatus();
         if (from == Service.State.STOPPING) {
           // Service was killed
+          runStatus = ProgramController.State.KILLED.getRunStatus();
+        }
+
+        if (workflowRunId == null) {
```

Comment:
Since systemArgs contain the Workflow related details, using them in the store to update the Workflow run accordingly. So that we only need the generic `setStart` and `setStop` methods.